accepted/2022/simplify-output-paths.md

```diff
@@ -0,0 +1,87 @@
+
+The logic for determining which output path format will be used will (mostly) be as follows:
+
+- If a project targets .NET 8 or higher, it will by default use the new output path format.  If it targets .NET 7 or lower, it will by default use the old output path format
+- A project can explicitly choose which output path format to use by setting the `UseArtifactsOutput` property to `true` or `false` in a `Directory.Build.props` file.
+
+### That pesky intermediate output path
+
+Unfortunately, the base intermediate output path can't depend on the project's `TargetFramework`, or anything else in the project file.  The only place to set properties which affect the base intermediate output path is in a `Directory.Build.props` file.  This is because MSBuild imports "project extensions" of the form `<ProjectName>.*.props` and `<ProjectName>.*.targets` from the base intermediate output path, and the project extensions .props files are imported early on in evaluation, before the body of the project file is evaluated.
+
+So, the way we will handle this is:
+
+- If `UseArtifactsOutput` is set to `true` in a `Directory.Build.props` file, then the new output format will be used, including for the intermediate output path
+- Otherwise, if `UseArtifactsOutput` is set to true later, either in the project file or in .NET SDK logic due to the `TargetFramework` being at least .NET 8, then the old output format will be used for the intermedate folder (so it will start with `obj\<Configuration>` by default), but the new `artifacts` format will be used for all other output.
+
+### Mixed output path formats in multi-targeted builds
+
+If a project is multi-targeted to `net7.0` and `net8.0` and doesn't specify the output path format, then different formats will be used for the output for the different target frameworks.  The .NET 7 build output will be in `bin\Debug\net7.0`, while the .NET 8 output will be in `artifacts\bin\debug_net8.0`.
+
+To prevent the output from one target framework to be globbed as part of the inputs to another target framework, both the `bin` and `artifacts` folder will need to be excluded from the default item globs (via the `DefaultItemExcludes` property).  This means that even projects that target .NET 7 or lower could be impacted by breaking changes if they currently have source files or other assets in an `artifacts` folder.
+
+Question: Could we have the outer build of a multi-targeted build query the inner builds to determine whether any of them target .NET 8 or higher, so we wouldn't have to have inconsistent output path formats?  Would this work with Visual Studio builds?
+
+### Solution or repo level output paths
```

Comment:
I'd strongly discourage this being an official feature. I know many people use it today and think they want it, but this pattern tends to lead to badness.

For CI builds it's common to use the repo-level output path as the drop root. So this would lead to all outputs from every project, including intermediates and UT outputs, to appear there. I've been preaching to everyone who will listen to just let the project-relative output dirs and only have your "runnables/shippables" do the aggregation or outputs/runtime assets and copy to the drop location.

Now you can work around some of this by having your drop root be `<root>\artifacts\bin`, but you still get all library output and UT outputs.

To be fair, this has always been a problem, but making this repo-level output path patterns be "officially well-supported" makes my discouragement of the practice carry less weight.

For reference, in our internal build system we're starting to see problems with teams using this practice as there are 1-2 million files in the drop leading to monstrous drops which are a pretty big waste of resources. Looking at one of these in particular recently, approx 1/3 was UT outputs (which don't need to ship...), and that's not counting whatever extraneous library outputs there are. I did a similar exercise in the VS repo recently and was able to reduce their drop by ~1 million files by cutting out UTs from the drop. So I don't think the concern for avoiding encouragement of this pattern is unwarranted.

Reply:
> For CI builds it's common to use the repo-level output path as the drop root

You mean because people would just blindly upload the whole `artifacts\` folder? I agree that is usually not what you want but I'd assume most people would upload `artifacts\packages` instead (or wherever you configure your runnables/shippables to end up in).

Reply:
> You mean because people would just blindly upload the whole artifacts\ folder?

Yes. That's exactly what people do today by manually redirecting the output paths. As mentioned, that's why we're seeing multiple instances internally from multiple orgs of drops with millions of extraneous files which aren't used in deployment. This would make that "officially" supported.

Reply:
@dfederm 

What's the problem except for excessive uploads? I'm of the opinion that if many people want a certain thing it's better to make this an official feature because it means the number of ways people achieve is limited which in turn often makes it viable to detect bad patterns and warning accordingly. For instance, uploading artifacts could warn when it includes known output types one usually doesn't want.

As far as development goes, I dislike the current structure with a passion because it makes local development and explicit cleaning more difficult. When you suspect an issue due to incremental builds, `build clean` is oftne not what you need. `git clean` usually doesn't work very well because it tries to delete IDE caches which are locked when the IDE is open. If you have a single folder in the solution/root level, deleting it and exploring it for troubleshooting is much much nicer. Of course, in the current system it's not that great because every project cooks up their own naming conventions and structures.

Reply:
Having at a solution level the ability to say all build outputs go in some other folder possibly outside of the source directory would be useful for some AWS tooling. We have tooling that mounts source directories as read only in a container and then do all of the building in a separate writable folder structure. It works well for many other languages but it is a pretty miserable experience for .NET given how much writing in the source directory happens during a build.